src/MatrixFields/var_dict.jl

```diff
@@ -0,0 +1,368 @@
+
+struct FieldMatrixStyle <: Base.Broadcast.BroadcastStyle end
+
+const FieldMatrixStyleType =
+    Union{FieldVectorViewBroadcasted, FieldMatrixBroadcasted}
+
+const FieldVectorStyleType = Union{
+    Fields.FieldVector,
+    Base.Broadcast.Broadcasted{<:Fields.FieldVectorStyle},
+}
+
+Base.Broadcast.broadcastable(vector_or_matrix::FieldMatrixStyleType) =
+    vector_or_matrix
+Base.Broadcast.broadcastable(vector::FieldVectorView) =
+    FieldVectorViewBroadcasted(keys(vector), values(vector))
+function Base.Broadcast.broadcastable(matrix::FieldMatrix)
+    entries = unrolled_map(values(matrix)) do entry
+        entry isa UniformScaling ? (entry,) : entry
+    end
+    return FieldMatrixBroadcasted(keys(matrix), entries)
+end
+
+is_wrapped_I(x) = x isa Tuple{UniformScaling}
+
+Base.Broadcast.BroadcastStyle(::Type{<:FieldMatrixStyleType}) =
+    FieldMatrixStyle()
+Base.Broadcast.BroadcastStyle(::FieldMatrixStyle, ::Fields.FieldVectorStyle) =
+    FieldMatrixStyle()
+
+function field_matrix_style_error(f, args)
+    arg_string(::Union{FieldVectorViewBroadcasted, FieldVectorStyleType}) =
+        "<vector>"
+    arg_string(::FieldMatrixBroadcasted) = "<matrix>"
+    arg_string(::T) where {T} = error(
+        "Unsupported FieldMatrixStyle broadcast argument type: $(T.name.name)",
+    )
+    args_string = join(map(arg_string, args), ", ")
+    error("Unsupported FieldMatrixStyle broadcast operation: $f.($args_string)")
+end
+
+Base.Broadcast.broadcasted(::FieldMatrixStyle, f, args...) =
+    field_matrix_style_error(f, args)
+
+# When a broadcast expression with + or * has more than two arguments, split it
+# up into a chain of two-argument broadcast expressions. This simplifies the
+# remaining methods for Base.Broadcast.broadcasted, since it allows us to assume
+# that they will have at most two arguments.
+Base.Broadcast.broadcasted(
+    ::FieldMatrixStyle,
+    f::Union{typeof(+), typeof(*)},
+    arg1,
+    arg2,
+    arg3,
+    args...,
+) =
+    unrolled_foldl((arg1, arg2, arg3, args...)) do arg1′, arg2′
+        Base.Broadcast.broadcasted(f, arg1′, arg2′)
+    end
+
+function Base.Broadcast.broadcasted(
+    ::FieldMatrixStyle,
+    ::typeof(-),
+    vector_or_matrix::FieldMatrixStyleType,
+)
+    VarDictType = var_dict_type(vector_or_matrix)
+    entries = unrolled_map(values(vector_or_matrix)) do entry
+        Base.Broadcast.broadcasted(-, entry)
+    end
+    return VarDictType(keys(vector_or_matrix), entries)
+end
+
+function Base.Broadcast.broadcasted(
+    ::FieldMatrixStyle,
+    f::Union{typeof(+), typeof(-)},
+    vector_or_matrix1::FieldMatrixStyleType,
+    vector_or_matrix2::FieldMatrixStyleType,
+)
+    var_dict_type(vector_or_matrix1) == var_dict_type(vector_or_matrix2) ||
+        field_matrix_style_error(f, (vector_or_matrix1, vector_or_matrix2))
+    VarDictType = var_dict_type(vector_or_matrix1)
+    all_keys = union(keys(vector_or_matrix1), keys(vector_or_matrix2))
```

Comment:
`union` necessarily allocates, so this is likely one source of allocations. If you actually need the unique set of these keys, then this likely will either allocate or be type unstable. This doesn't seem ideal.

Reply:
I wrote a custom `union` (and `intersect`, `setdiff`, etc.) for `VarSet` that is type-stable and handles variable name overlaps.